tensorflow-framework/src/main/java/org/tensorflow/framework/activations/ReLU.java

```diff
@@ -0,0 +1,143 @@
+ * element-wise maximum of 0 and the input tensor.
+ *
+ * <p>Modifying default parameters allows you to use non-zero thresholds, change the max value of
+ * the activation, and to use a non-zero multiple of the input for values below the threshold.
+ *
+ * <p>For example:
+ *
+ * <pre>
+ *     Operand&lt;TFloat32&gt; input = tf.constant(
+ *              new float[] {-10f, -5f, 0.0f, 5f, 10f});
+ *
+ *     // With default parameters
+ *     ReLU&lt;TFloat32&gt; relu = new ReLU&lt;&gt;(tf);
+ *     Operand&lt;TFloat32&gt; result = relu.call(input);
+ *     // result is [0.f,  0.f,  0.f,  5.f, 10.f]
+ *
+ *     // With alpha = 0.5
+ *     relu = new ReLU&lt;&gt;(tf, 0.5f, ReLU.MAX_VALUE_DEFAULT, ReLU.THRESHOLD_DEFAULT);
+ *     result = relu.call(input);
+ *     // result is [-5.f , -2.5f,  0.f ,  5.f , 10.f]
+ *
+ *     // With maxValue = 5
+ *     relu = new ReLU&lt;&gt;(tf, ReLU.ALPHA_DEFAULT, 5f, ReLU.THRESHOLD_DEFAULT);
+ *     result = relu.call(input);
+ *     // result is [0.f, 0.f, 0.f, 5.f, 5.f]
+ *
+ *     // With threshold = 5
+ *     relu = new ReLU&lt;&gt;(tf, ReLU.ALPHA_DEFAULT, ReLU.MAX_VALUE_DEFAULT, 5f);
+ *     result = relu.call(input);
+ *     // result is [-0.f, -0.f,  0.f,  0.f, 10.f]
+ * </pre>
+ *
+ * @param <T> the data type of the result
+ */
+public class ReLU<T extends TNumber> extends Activation<T> {
+
+  public static final float ALPHA_DEFAULT = 0.0f;
+  public static final float MAX_VALUE_DEFAULT = Float.NaN;
+  public static final float THRESHOLD_DEFAULT = 0.0f;
+
+  private final float alpha;
+  private final float maxValue;
+  private final float threshold;
+
+  /**
+   * Creates a new ReLU with alpha={@link #ALPHA_DEFAULT}, maxValue={@link #MAX_VALUE_DEFAULT},
+   * threshold={@link #THRESHOLD_DEFAULT},
+   *
+   * @param tf the TensorFlow Ops
+   */
+  public ReLU(Ops tf) {
+    this(tf, ALPHA_DEFAULT, MAX_VALUE_DEFAULT, THRESHOLD_DEFAULT);
+  }
+
+  /**
+   * Creates a new ReLU
+   *
+   * @param tf the TensorFlow Ops
+   * @param alpha governs the slope for values lower than the threshold.
+   * @param maxValue sets the saturation threshold (the largest value the function will return).
+   * @param threshold the threshold value of the activation function below which values will be
+   *     damped or set to zero.
+   */
+  public ReLU(Ops tf, float alpha, float maxValue, float threshold) {
+    super(tf);
+    this.alpha = alpha;
+    this.maxValue = maxValue;
+    this.threshold = threshold;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Operand<T> call(Operand<T> input) {
+
+    DataType<T> dataType = input.asOutput().dataType();
+    Operand<T> negativePart = null;
+    boolean clipMax = !Float.isNaN(maxValue);
+    if (alpha != 0) {
+      if (Float.isNaN(maxValue) && threshold == 0) {
+        // TODO LeakyRelu is not in tf.nn ????
+        return LeakyRelu.create(tf.scope(), input, LeakyRelu.alpha(alpha));
+      }
+      if (threshold != 0) {
+        negativePart =
```

Comment:
What would you think about moving both the declaration and the initialization of `negativePart` down to immediately before it is used? I realize that would require a change to keep `input` as the original input instead of reusing it for intermediate calculations, but personally I'd see that as a feature not a bug.

Reply:
Perhaps rename `negativePart` -> `amountBelowThreshold`? For one thing, `negativePart` is vague to me. For another, it's always positive.

Reply:
I am not sure if this is used here as a math term, but I found [Wolfram MathWorld - Negative Part](https://mathworld.wolfram.com/NegativePart.html). Perhaps @Craigacp  and @karllessard  might want to comment.

Reply:
Good point, that Wolfram definition does show why `negativePart` was used in the first place. That said, I'd still advocate that in the presence of `threshold`, it's a confusing name.